--- a/linux/include/openswan/ipsec_kversion.h
+++ b/linux/include/openswan/ipsec_kversion.h
@@ -614,6 +614,15 @@
 # define DEFINE_RWLOCK(x) rwlock_t x = RW_LOCK_UNLOCKED
 #endif
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,15,10)
+# define IP_SELECT_IDENT_NEW2
+#elif LINUX_VERSION_CODE < KERNEL_VERSION(3,15,0) && LINUX_VERSION_CODE >= KERNEL_VERSION(3,14,17)
+# define IP_SELECT_IDENT_NEW2
+#elif LINUX_VERSION_CODE < KERNEL_VERSION(3,11,0) && LINUX_VERSION_CODE >= KERNEL_VERSION(3,10,53)
+# define IP_SELECT_IDENT_NEW2
+#elif LINUX_VERSION_CODE < KERNEL_VERSION(3,5,0) && LINUX_VERSION_CODE >= KERNEL_VERSION(3,4,103)
+# define IP_SELECT_IDENT_NEW2
+#endif
 
 #endif /* _OPENSWAN_KVERSIONS_H */
 
--- a/linux/include/openswan/ipsec_param2.h
+++ b/linux/include/openswan/ipsec_param2.h
@@ -48,6 +48,11 @@
 #define KLIPS_IP_SELECT_IDENT(iph, skb) ip_select_ident(iph, skb_dst(skb), NULL)
 #endif
 
+/* kernels > 2.4.102 */
+#if defined(IP_SELECT_IDENT) && defined(IP_SELECT_IDENT_NEW2)
+#define KLIPS_IP_SELECT_IDENT(iph, skb) ip_select_ident(skb, NULL)
+#endif
+
 #endif /* SUSE_LINUX_2_4_19_IS_STUPID */
 
 #if !defined(HAVE_CURRENT_UID)
