--- a/pppd/ipcp.c
+++ b/pppd/ipcp.c
@@ -1777,9 +1777,9 @@ ip_demand_conf(u)
 	if (sifproxyarp(u, wo->hisaddr))
 	    proxy_arp_set[u] = 1;
 
-    notice("%s : pid : %d : local  IP address %I",ifname,getpid(),wo->ouraddr);
+    notice("local  IP address %I",wo->ouraddr);
     if (wo->hisaddr)
-	notice("remote IP address %I", wo->hisaddr);
+	notice("%s : pid : %d : remote IP address %I",ifname,getpid(),wo->hisaddr);
 
     return 1;
 }
@@ -1947,9 +1947,9 @@ ipcp_up(f)
 
 	ipcp_wantoptions[0].ouraddr = go->ouraddr;
 
-	notice("%s : pid : %d :local  IP address %I", ifname,getpid(),go->ouraddr);
+	notice("local  IP address %I", go->ouraddr);
 	if (ho->hisaddr != 0)
-	    notice("remote IP address %I", ho->hisaddr);
+	    notice("%s : pid : %d : remote IP address %I", ifname,getpid(),ho->hisaddr);
 	if (go->dnsaddr[0])
 	    notice("primary   DNS address %I", go->dnsaddr[0]);
 	if (go->dnsaddr[1])
